### ISSUE TITLE
Audit personalization_gain metric calculation for correctness

### ISSUE THREAD

**Issue by reinesaj2** [reinesaj2/federated-ids#164]
## Problem Statement

Related to Issue #163: Need to verify that `personalization_gain` metric is calculated correctly and measures what we intend.

## Current Calculation (suspected location: client.py)

```python
# Hypothesized calculation:
personalization_gain = personalized_model_f1 - global_model_f1
```

### 1. Same Test Set Used for Both Models
- If global and personalized models evaluated on **same test data**, gain could be artificially low
- **Expected**: Global model tested on aggregated test set, personalized on client's local test set
- **Verify**: Confirm test sets are properly isolated

### 2. Model Identity Bug
- If `personalized_model == global_model` (no actual fine-tuning), gain = 0.0
- **Evidence from Issue #163**: Median gain = 0.0 across 1,600 clients
- **Verify**: Assert models differ after personalization phase

### 3. Incorrect Baseline
- Gain should compare **same client's** performance before/after personalization
- **Incorrect**: `gain = client_personalized_f1 - mean(all_clients_global_f1)`
- **Correct**: `gain = client_personalized_f1 - client_global_f1`

### 4. Evaluation Mode Not Set
```python
# Bug: Model not in eval mode during testing
model.train()  # ❌ Dropout/BatchNorm in training mode
y_pred = model(X_test)

# Correct:
model.eval()  # ✅ Deterministic inference
with torch.no_grad():
    y_pred = model(X_test)
```

### Code Audit:
- [ ] Review `client.py` personalization evaluation logic
- [ ] Trace data flow: which test set is used for each model
- [ ] Verify model checkpointing: global vs personalized are distinct files
- [ ] Check torch determinism: eval mode, no_grad context

### Unit Tests to Add:
```python
def test_personalization_gain_calculation():
    """Verify gain metric computes correctly."""
    # Mock: personalized model performs 5% better than global
    global_f1 = 0.80
    personalized_f1 = 0.84
    
    gain = compute_personalization_gain(global_f1, personalized_f1)
    assert gain == 0.04, f"Expected 0.04, got {gain}"
    assert 0 < gain < 1, "Gain should be bounded"

def test_personalization_changes_model():
    """Ensure fine-tuning actually modifies weights."""
    global_model = load_global_model()
    personalized_model = personalize(global_model, local_data)
    
    # Weights MUST differ after fine-tuning
    assert not torch.equal(
        global_model.state_dict()['fc.weight'],
        personalized_model.state_dict()['fc.weight']
    ), "Personalization did not change model weights!"
```

### Diagnostic Experiment:
Run controlled test where personalization **must** help:
```bash
# Client with 100% attack traffic, global model trained on 100% benign
# Personalized model MUST adapt and show positive gain
python scripts/test_personalization_sanity.py \
  --client_data=attack_only \
  --global_data=benign_only \
  --personalization_epochs=5
```

**Expected**: `gain > 0.3` (30% improvement)  
**If gain ≈ 0**: Metric calculation is broken

## Acceptance Criteria

- [ ] Gain calculation audited and documented in code comments
- [ ] Unit tests added for metric edge cases (negative gain, >1.0 gain, NaN handling)
- [ ] Sanity check experiment shows positive gains in controlled scenario
- [ ] Issue #163 updated with findings (either "metric correct, implementation broken" or "metric bug found")

## Deliverables

1. **Audit Report** (markdown doc):
   - Current implementation trace
   - Identified bugs or correct-as-is confirmation
   - Recommendations for fixes

2. **Test Suite** (`test_personalization_metrics.py`):
   - Edge case coverage
   - Property-based tests (gains must be bounded, symmetric)

3. **CI Integration**:
   - Add metric validation to personalization workflow
   - Fail if gains are all zero (catch regression)

## Related Issues

- Issue #163 (personalization producing zero gains - this may be root cause)
- Issue #43 (original investigation - closed prematurely)

## Priority

**P1**: Must confirm metric correctness before debugging implementation (Issue #163). If metric is broken, fixing implementation won't help.
